source/agora/consensus/validation/Block.d

```diff
@@ -73,8 +80,7 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
 
     // special case for the genesis block
     if (block.header.height == 0)
-        return block == GenesisBlock ?
-            null : "Block: Height 0 but not Genesis block";
+        return isValidGenesisBlock(block);
```

Comment:
Alright, I see there is actually a problem with this now

There is a possibility of a DDoS here. We should only accept a block with height zero if none was stored yet. If we already have a block with height 0 in our ledger, then calling `isValidGenesisBlock` on blocks with height zero received from the network becomes a form of DDoS. We should outright reject such blocks.

We only set a custom genesis block through the Config file. Therefore we never have to verify blocks with height 0 coming from the network. They should be rejected.

Here is what I suggest:

Inside here simply do:

```
if (block.header.height == 0)
    return "Block: Rejected Genesis block"
```

And you will indeed have to make `isValidGenesisBlock` `public`.

Furthermore, there is a critical piece of code missing from your PR. `setGenesisBlock` is not validating the genesis block which was set. This will have to be fixed to call this new `public` function.

Reply:
That's right. It's an important point.

And we can setGenesisBlock only in FullNode.